InfrastructureLayer/Logging/Logging.cpp

```diff
@@ -0,0 +1,190 @@
+#include "Logging.h"
+
+#include <QDebug>
+#include <QFile>
+#include <QDate>
+
+Logging* Logging::instance_ = new Logging;
+
+namespace
+{
+    QString LOG_DIR = "log/";
+    QString LOG_NAME = "epsilon-hermes-";
+    QString LOG_DATE_FORMAT = "yyyy-MM-dd";
+    QString LOG_EXT = ".log";
+
+    QString LOG_DATETIME_FORMAT = "yyyy-MM-dd-hh-mm-ss-zzz-t";
+    QString DEBUG_MSG_FORMAT = "[%1] (%2, %3:%4) %5\n"; // datetime, file, function, line, msg
+    QString NON_DEBUG_MSG_FORMAT = "[%1] (%2) %3\n"; // datetime, file, msg
+}
+
+void Logging::init(int level)
+{
+    qInstallMessageHandler(Logging::logMessageHandler);
+
+    switch (level)
+    {
+        case 0:
+            logLevel_ = Logging::DEBUG;
+            break;
+
+        case 1:
+            logLevel_ = Logging::INFO;
+            break;
+
+        case 2:
+            logLevel_ = Logging::WARNING;
+            break;
+
+        case 3:
+            logLevel_ = Logging::CRITICAL;
+            break;
+
+        case 4:
+            logLevel_ = Logging::FATAL;
+            break;
+
+        default:
+            logLevel_ = Logging::DEBUG;
+            qWarning() << "Configured log level unrecognized. "
+                       << "DEBUG=0,INFO=1,WARNING=2,CRITICAL=3,FATAL=4. "
+                       << "Defaulting to DEBUG.";
+            break;
+    }
+
+    // Because Epsilon Hermes is meant for a solar car
+    // and competitions do not include night racing
+    // It is quite unlikely to for the date to change while the car is running
+    // So this assumes the date will not change while Epsilon Hermes is running
+    QString todayStr = QDate::currentDate().toString(LOG_DATE_FORMAT);
+    QString logName = LOG_DIR + LOG_NAME + todayStr + LOG_EXT;
+    logFile_.setFileName(logName);
+
+    if (logFile_.open(QIODevice::WriteOnly | QIODevice::Append))
+    {
+        logStream_.setDevice(&logFile_);
+        qInfo() << "Logging Initialized";
+    }
+    else
+    {
+        qFatal("Logging initalization failed");
+    }
+
+}
+
+QTextStream& Logging::logStream()
+{
+    return logStream_;
+}
+
+Logging::LogLevel Logging::logLevel()
+{
+    return logLevel_;
+}
+
+Logging* Logging::instance()
```

Comment:
This should return a reference, returning a pointer gives the class user a temptation that it could manage the memory.

This implementation of the singleton is also leaking memory but what could be a bit more significant of a problem is that the destructor is not called which means the QFile destructor is not called and the file is never closed. I prefer the  static variable way of 
```c++
Logging& Logging::instance() 
{
   static Logging instance;
   return instance;
}
```
See https://stackoverflow.com/a/32679888

Reply:
doh